File: solutions/php/01-cq2/code/app/main.php
<?php
error_reporting(E_ALL);

function match_pattern($input_line, $pattern) {
    if (strlen($pattern) == 1) {
        return str_contains($input_line, $pattern);
    } else {
        throw new RuntimeException("Unhandled pattern: $pattern");
    }
}

function main($argv) {
    $pattern = $argv[2];
    $input_line = stream_get_contents(STDIN);

    if ($argv[1] !== "-E") {
        fwrite(STDERR, "Expected first argument to be '-E'\n");
        exit(1);
    }

    if (match_pattern($input_line, $pattern)) {
        exit(0);
    } else {
        exit(1);
    }
}

main($argv);

File: solutions/c/01-cq2/code/src/main.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

bool match_pattern(const char* input_line, const char* pattern) {
    if (strlen(pattern) == 1) {
        return strchr(input_line, pattern[0]) != NULL;
    } else {
        fprintf(stderr, "Unhandled pattern %s\n", pattern);
        exit(1);
    }
}

int main(int argc, char* argv[]) {
    // Disable output buffering
    setbuf(stdout, NULL);
    setbuf(stderr, NULL);

    if (argc != 3) {
        fprintf(stderr, "Expected two arguments\n");
        return 1;
    }

    const char* flag = argv[1];
    const char* pattern = argv[2];

    if (strcmp(flag, "-E") != 0) {
        fprintf(stderr, "Expected first argument to be '-E'\n");
        return 1;
    }

    char input_line[1024];
    if (fgets(input_line, sizeof(input_line), stdin) == NULL) {
        return 1;
    }

    // Remove trailing newline
    input_line[strcspn(input_line, "\n")] = '\0';

    if (match_pattern(input_line, pattern)) {
        return 0;
    } else {
        return 1;
    }
}
